providers/implementations/ciphers/cipher_aes_cts_fips.c

```diff
@@ -0,0 +1,313 @@
+
+ * AES_CBC_CTS has 3 variants:
+ *  (1) CS1 The NIST variant.
+ *      If the length is a multiple of the blocksize it is the same as CBC mode.
+ *      otherwise it produces C1||C2||(C(n-1))*||Cn.
+ *      Where C(n-1)* is a partial block.
+ *  (2) CS2
+ *      If the length is a multiple of the blocksize it is the same as CBC mode.
+ *      otherwise it produces C1||C2||Cn||(C(n-1))*.
+ *      Where C(n-1)* is a partial block.
+ *  (3) CS3 The Kerberos5 variant.
+ *      Produces C1||C2||Cn||(C(n-1))* regardless of the length.
+ *      If the length is a multiple of the blocksize it looks similiar to CBC mode
+ *      with the last 2 blocks swapped.
+ *      Otherwise it is the same as CS2.
+ */
+
+#include "e_os.h" /* strcasecmp */
+#include <openssl/core_names.h>
+#include "prov/ciphercommon.h"
+#include "internal/nelem.h"
+#include "cipher_aes_cts.h"
+
+/* The value assigned to 0 is the default */
+#define CTS_CS1 0
+#define CTS_CS2 1
+#define CTS_CS3 2
+
+typedef union {
+    size_t align;
+    unsigned char c[16];
+} aligned_16bytes;
+
+typedef union {
+    size_t align;
+    unsigned char c[32];
+} aligned_32bytes;
+
+typedef struct cts_mode_name2id_st {
+    unsigned int id;
+    const char *name;
+} CTS_MODE_NAME2ID;
+
+static CTS_MODE_NAME2ID cts_modes[] =
+{
+    { CTS_CS1, OSSL_CIPHER_CTS_MODE_CS1 },
+#ifndef FIPS_MODULE
+    { CTS_CS2, OSSL_CIPHER_CTS_MODE_CS2 },
+    { CTS_CS3, OSSL_CIPHER_CTS_MODE_CS3 },
+#endif
+};
+
+const char *aes_cbc_cts_mode_id2name(unsigned int id)
+{
+    size_t i;
+
+    for (i = 0; i < OSSL_NELEM(cts_modes); ++i) {
+        if (cts_modes[i].id == id)
+            return cts_modes[i].name;
+    }
+    return NULL;
+}
+
+int aes_cbc_cts_mode_name2id(const char *name)
+{
+    size_t i;
+
+    for (i = 0; i < OSSL_NELEM(cts_modes); ++i) {
+        if (strcasecmp(name, cts_modes[i].name) == 0)
+            return (int)cts_modes[i].id;
+    }
+    return -1;
+}
+
+static size_t cts128_cs1_encrypt(PROV_CIPHER_CTX *ctx, const unsigned char *in,
+                                 unsigned char *out, size_t len)
+{
+    aligned_16bytes tmp_in;
+    size_t residue;
+
+    residue = len % 16;
+    len -= residue;
+    if (!ctx->hw->cipher(ctx, out, in, len))
+        return 0;
+
+    if (residue == 0)
+        return len;
+
+    in += len;
+    out += len;
+
+    memset(tmp_in.c, 0, sizeof(tmp_in));
+    memcpy(tmp_in.c, in, residue);
+    if (!ctx->hw->cipher(ctx, out - 16 + residue, tmp_in.c, 16))
+        return 0;
+    return len + residue;
+}
+
+static size_t cts128_cs1_decrypt(PROV_CIPHER_CTX *ctx, const unsigned char *in,
+                                 unsigned char *out, size_t len)
+{
+    aligned_32bytes tmp_out;
+    size_t residue;
+
+    residue = len % 16;
+    if (residue == 0) {
+        /* If there are no partial blocks then it is the same as CBC mode */
+        if (!ctx->hw->cipher(ctx, out, in, len))
+            return 0;
+        return len;
+    }
+    len -= 16 + residue;
```

Comment:
Sanity check that `len` does not  underflow? You have this check for CS3 but not here?

Reply:
Not quite sure how I lost that line.. (Added to encrypt also)

Reply:
Ahh I remember now - the parent caller aes_cbc_cts_block_update() does the check.
CS3 needs an extra check because it needs 2 blocks Cs1 & Cs2 only need 'at least' one block.

Reply:
Will take it back out again where I just added it- also added failure test cases for these.